doc_src/bind.txt

```diff
@@ -131,3 +131,14 @@ set -g fish_key_bindings fish_vi_key_bindings
 bind -M insert \cc kill-whole-line force-repaint
 \endfish
 Turns on Vi key bindings and rebinds @key{Control,C} to clear the input line.
+
+
+\subsection special-case-escape Special Case: The escape Character
+
+The escape key (or character) poses a special challenge for fish. Consider Vi mode. On the one hand you want to be able to press the escape key to switch from insert to normal mode. On the other hand you want fish to recognize multi-char sequences that begin with an escape character (which is what the function and arrow-keys send) without having the escape character switch to normal mode. So fish needs to wait a little bit to see if the escape introduces a multi-char sequence but not so long that you're left wondering if fish will ever switch to Vi normal mode after you press escape.
```

Comment:
Even though it is not unprecedented in Fish's command reference to use direct speech ("you"), I would consider rephrasing the majority of this text to use indirect speech. It sounds more appropriate to me (maybe I'm too old fashioned? what do you think?).

Following is a proposal with mostly indirect speech. This is actually more than slight rephrasing, but it's not as if to say that I think it's better now than before. My text is simply different (and worse, if anything).

> Since the escape key (or character) plays two distinct roles, the escape key poses a special challenge for fish. In its first role, it stands by itself. In Vi mode, for example, escape is used to switch from insert to normal mode. In its second role, however, escape is used as a "meta" key and then it's only the beginning of some longer key-sequence. Coming back to the Vi mode example, sometimes fish is expected to realize that the escape key should be regarded as "meta", meaning that the escape character is part of some multi-char sequence. Function keys (F1, F2, etc...) and arrow keys are common cases of multi-char sequences sent with the escape character as "meta", but escape can be used in this manner to define custom bindings as well. Obviously, fish is not supposed to exit insert mode when the escape is part of a longer key-sequence.

> To be able distinguish between these two roles, fish has to wait after it sees an escape character. In this waiting period, any additional key presses make the escape key behave as "meta". Otherwise, it remains simply an escape key press. The waiting period is set to 500 milliseconds (0.5 seconds) by default This is the Gnu readline library default escape timeout. It can be configured by setting the `fish_escape_delay_ms` variable to a value between 10 and 5000 ms. It is recommended that this be a universal variable that you set once from an interactive session.

(The note on previous versions remains unchanged)

The rest of the PR looks great to me, excluding test code, which I am not familiar with and didn't have time to go over.